openscm/core.py

```diff
@@ -1,24 +1,304 @@
+        res = self._children.get(name, None)
+        if res is None:
+            if self._has_been_written_to:
+                raise ParameterWrittenError
+            if self._has_been_aggregated:
+                raise ParameterAggregatedError
+            res = _Parameter(name)
+            res._parent = self
+            res._type = parameter_type
+            res._unit = unit
+            self._children[name] = res
+        return res
+
+    def attempt_aggregate(self, parameter_type: ParameterType) -> None:
+        """
+        Tell parameter that it will be read from in an aggregated way, i.e., aggregating
+        over child parameters.
+
+        Parameters
+        ----------
+        parameter_type
+            Parameter type to be read
+
+        Raises
+        ------
+        ParameterTypeError
+            If parameter has already been read from or written to in a different type.
+        """
+        if self._type is not None and self._type != parameter_type:
+            raise ParameterTypeError
+        self._type = parameter_type
+        self._has_been_aggregated = True
+
+    def attempt_write(self, parameter_type: ParameterType) -> None:
+        """
+        Tell parameter that its data will be written to.
+
+        Parameters
+        ----------
+        parameter_type
+            Parameter type to be written
+
+        Raises
+        ------
+        ParameterReadonlyError
+            If parameter is read-only because it has child parameters.
+        """
+        if len(self._children) > 0:
+            raise ParameterReadonlyError
+        self.attempt_aggregate(parameter_type)
+        self._has_been_written_to = True
+
+    @property
+    def full_name(self) -> Tuple[str]:
+        """
+        Full hierarchical name
+        """
+        p = self
+        r = []
+        while p is not None:
+            r.append(p.name)
+            p = p._parent
+        return tuple(reversed(r))
+
+    @property
+    def name(self) -> str:
+        """
+        Name
+        """
+        return self._name
+
+    @property
+    def parameter_type(self) -> ParameterType:
+        """
+        Parameter type
+        """
+        return self._type
+
+    @property
+    def parent(self) -> "_Parameter":
+        """
+        Parent parameter
+        """
+        return self._parent
+
+    @property
+    def unit(self) -> str:
+        """
+        Unit
+        """
+        return self._unit
+
+
+class _Region:
+    """
+    Represents a region in the region hierarchy.
+    """
+
+    _children: Dict[str, "_Region"]
+    """Subregions"""
+
+    _has_been_aggregated: bool
+    """
+    Tells if a parameter of this region has already been read in an aggregated way,
+    i.e., aggregating over subregions
+     """
+
+    _name: str
+    """Name"""
+
+    _parameters: Dict[str, _Parameter]
+    """Parameters"""
+
+    _parent: "_Region"
+    """Parent region (or `None` if root region)"""
+
+    def __init__(self, name: str):
```

Comment:
can we somehow avoid duplicating `__init__` between `Region` and `Parameter`? Superclass?

Reply:
We could in principle, but then the typing of `_parent` and `_children` would not be really correct (we would need generic programming abilities here...). Let's use these for now and the refactor once the whole thing works, ok?

Reply:
good plan. we may need to use different words other than parent and child given their pre-existing special meanings

Reply:
As I perceive it, these words are rather common when programming hierarchical structures. I would, personally like to stick to them.